File: sirent/player.py
import discord
from sirent.fs import *
import subprocess
import os
import security_p

__root = os.path.dirname(os.getcwd())
v_obj = None


def init(client):
    @client.command(name="leave")
    async def _leave(ctx):
        if not security_p.has_permission(ctx):
            return
        global v_obj
        write_cmd("leave")
        write_log("info", "Leave channel")
        await ctx.channel.send("Leave channel")
        await v_obj.disconnect()

    @client.command(name="play")
    async def _play(ctx, url=""):
        if not security_p.has_permission(ctx):
            return
        global v_obj
        write_cmd("play")
        if url == "":
            write_log("error", "The argument url is not given")
            return
        write_log("info", f"Play from url: {url}")
        subprocess.call(os.path.dirname(os.getcwd()) + "\\assets\\ytdl\\ytdl.exe " +
                        url + " " + os.path.dirname(os.getcwd()) + "\\assets\\music\\music.wav")
        while True:
            if os.path.exists("../assets/music/music.wav"):
                break

        channel = ctx.author.voice.channel
        v_obj = await channel.connect()
        if v_obj is not None:
            if v_obj.is_playing():
                v_obj.stop()
        v_obj.play(discord.FFmpegPCMAudio(executable=__root + "\\assets\\ffmpeg\\ffmpeg.exe",
                                          source=__root + "\\assets\\music\\music.wav"
                                          ))
        v_obj.source = discord.PCMVolumeTransformer(v_obj.source)
        v_obj.source.volume = 0.5

    @client.command(name="pause")
    async def _pause(ctx):
        if not security_p.has_permission(ctx):
            return
        write_cmd("pause")
        global v_obj
        if not v_obj:
            write_log("error", "Can not pause because no url is playing")
            ctx.channel.send("Can not pause because no url is playing")
            return

        write_log("error", "Pausing the played url")
        ctx.channel.send("Pausing the music")
        v_obj.pause()

    @client.command(name="resume")
    async def _resume(ctx):
        if not security_p.has_permission(ctx):
            return
        global v_obj
        if not v_obj:
            write_log("error", "Can not resume playing the music because not url is playing")
            ctx.channel.send("Can not resume playing the music because not url is playing")
            return
        v_obj.resume()

File: sirent/player.ts
import {Message, StreamDispatcher} from "discord.js";
import ytdl from "ytdl-core";
import path from "path";
import * as fs from "fs";

export class MusicPlayer {
    private player: StreamDispatcher | undefined;
    private playlist: string[];
    private playlistStep: number;
    private playlistLoop: boolean;

    constructor() {
        this.playlist = [];
        this.playlistStep = -1;
        this.playlistLoop = false;
    }

    public async setLoop(loop: boolean): Promise<void> {
        this.playlistLoop = loop;
    }

    public async getLoop(): Promise<boolean> {
        return this.playlistLoop;
    }

    public async addPlaylistItem(ctx: Message, url: string): Promise<void> {
        if (!ctx.member?.voice.channel) {
            await ctx.channel.send("Please join a voice channel");
            return;
        }
        let matches = url.match(/watch\?v=([a-zA-Z0-9\-_]+)/);
        if (matches) {
            this.playlist.push(url);
        } else {
            await ctx.channel.send("Your URL is invalid");
        }
    }

    public async play(ctx: Message, url: string): Promise<void> {
        if (!ctx.member?.voice.channel) {
            await ctx.channel.send("Please join a voice channel");
            return;
        }
        let matches = url.match(/watch\?v=([a-zA-Z0-9\-_]+)/);
        if (matches) {
            const connection = await ctx.member.voice.channel.join();
            this.player = connection.play(ytdl(url, {
                filter: "audioonly"
            }), {
                volume: 0.5
            });

            this.player.on("finish",() => {
                if (this.playlistStep === this.playlist.length -1) {
                    if (this.playlistLoop) {
                        this.playlistStep = 0;
                        this.play(ctx, this.playlist[this.playlistStep]);
                    }
                } else {
                    this.playlistStep++;
                    this.play(ctx, this.playlist[this.playlistStep]);
                }
            });
        } else {
            await ctx.channel.send("Your URL is invalid");
        }
    }

    public async pause(ctx: Message): Promise<void> {
        if (this.player !== undefined) {
            await ctx.channel.send("Pausing the playing music");
            this.player.pause();
        } else {
            await ctx.channel.send("No song is playing");
        }
    }
}
